include/fmt/chrono.h

```diff
@@ -966,13 +966,122 @@ inline void tzset_once() {
+// microseconds precision.
+template <long long Num, long long Den, int N = 0,
+          bool Enabled = (N < 19) && (Num <= max_value<long long>() / 10)>
+struct count_fractional_digits {
+  static constexpr int value =
+      Num % Den == 0 ? N : count_fractional_digits<Num * 10, Den, N + 1>::value;
+};
+
+// Base case that doesn't instantiate any more templates
+// in order to avoid overflow.
+template <long long Num, long long Den, int N>
+struct count_fractional_digits<Num, Den, N, false> {
+  static constexpr int value = (Num % Den == 0) ? N : 6;
+};
+
+// Format non-fractional subseconds with an appriate number of digits.
+template <typename OutputIt, typename Duration, typename char_type>
+void write_sub_sec(OutputIt& out, Duration d) {
+  FMT_ASSERT(!std::is_floating_point<typename Duration::rep>::value, "");
+  constexpr auto num_fractional_digits =
+      count_fractional_digits<Duration::period::num,
+                              Duration::period::den>::value;
+
+  using subsecond_precision = std::chrono::duration<
+      typename std::common_type<typename Duration::rep,
+                                std::chrono::seconds::rep>::type,
+      std::ratio<1, detail::pow10(num_fractional_digits)>>;
+  if (std::ratio_less<typename subsecond_precision::period,
+                      std::chrono::seconds::period>::value) {
+    *out++ = '.';
+    auto fractional =
+        detail::abs(d) - std::chrono::duration_cast<std::chrono::seconds>(d);
+    auto subseconds =
+        std::chrono::treat_as_floating_point<
+            typename subsecond_precision::rep>::value
+            ? fractional.count()
+            : std::chrono::duration_cast<subsecond_precision>(fractional)
+                  .count();
+    uint32_or_64_or_128_t<long long> n =
+        to_unsigned(to_nonnegative_int(subseconds, max_value<long long>()));
+    int num_digits = detail::count_digits(n);
+    if (num_fractional_digits > num_digits)
+      out = std::fill_n(out, num_fractional_digits - num_digits, '0');
+    out = format_decimal<char_type>(out, n, num_digits).end;
+  }
+}
+
+// Format fractional seconds with an appropiate number of digits.
+template <typename Rep, long long Num, long long Den>
+void format_sec_fractional(basic_memory_buffer<char>& buf, Rep val) {
+  FMT_ASSERT(std::is_floating_point<Rep>::value, "");
+  auto num_fractional_digits = count_fractional_digits<Num, Den>::value;
+  // For non-integer values, we ensure at least 6 digits to get microsecond
+  // precision.
+  if (num_fractional_digits < 6 && static_cast<Rep>(std::round(val)) != val) {
+    num_fractional_digits = 6;
+  }
+  format_to(std::back_inserter(buf), runtime("{:.{}f}"),
+            std::fmod(val * static_cast<Rep>(Num) / static_cast<Rep>(Den),
+                      static_cast<Rep>(60)),
+            num_fractional_digits);
+}
+
+template <typename OutputIt, typename Char, typename Rep = int,
+          typename Period = std::ratio<1, 1>>
```

Comment:
Let's replace Rep and Period with Duration for simplicity

Reply:
ok, also replaced in
```
template <typename FormatContext, typename Duration>
  auto format(const std::tm& tm, FormatContext& ctx,
              const Duration& subsecs) const
```